File: external_samples/servo.py
# @license
# Copyright 2025 Porpoiseful LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# @fileoverview This is a sample for a servo
# @author alan@porpoiseful.com (Alan Smith)

from component import Component, PortType, InvalidPortException

class Servo(Component):    
    def __init__(self, ports : list[tuple[PortType, int]]):
        portType, port = ports[0]
        if portType != PortType.SERVO_PORT:
            raise InvalidPortException
        self.port = port
    def get_manufacturer(self) -> str:   
        return "REV Robotics" 
    def get_name(self) -> str:  
        return "SRS Servo"
    def get_part_number(self) -> str:  
        return "REV-41-1097"
    def get_url(self) -> str:  
        return "https://www.revrobotics.com/rev-41-1097/"
    def get_version(self) -> tuple[int, int, int]: 
        return (1, 0, 0)
    def stop(self) -> None:
        # De-energize servo port
        pass
    def reset(self) -> None:
        pass
    def get_connection_port_type(self) -> list[PortType]:
        return [PortType.SERVO_PORT]
    def periodic(self) -> None:
        pass

    # Component specific methods
    def set_position(self, pos: float) -> None:
        '''Set the servo to a position between 0 and 1'''
        # sends to the hardware the position of the servo
        pass
    def set_angle_degrees(self, angle: float) -> None:
        '''Set the servo to an angle between 0 and 270'''
        self.set_position(angle / 270.0)
    
    
    

File: src/toolbox/hardware_components/servo.ts
/**
 * @license
 * Copyright 2025 Porpoiseful LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * @fileoverview Servo hardware component definitions.
 * @author alan@porpoiseful.com (Alan Smith)
 */

import * as Blockly from 'blockly/core';

import * as ToolboxItems from '../items';

export const TYPE_NAME = 'Servo';

/**
 * Returns a component definition block for a servo.
 */
export function getDefinitionBlock(hideParams: boolean): ToolboxItems.ContentsType {
  return {
    kind: 'block',
    type: 'mrc_component',
    fields: {
      NAME: 'my_servo',
      TYPE: 'RevServo',
    },
    extraState: {
      importModule: 'rev_servo',
      params: [{name: 'servo_port', type: 'int'}],
      hideParams,
    },
    ...(hideParams ? {} : {
      inputs: {
        ARG0: {
          shadow: {
            type: 'mrc_port',
            fields: {
              TYPE: 'SmartIO',
              PORT_NUM: 1,
            },
          },
        },
      },
    }),
  };
}

/**
 * Returns the blocks available for a servo component.
 */
export function getBlocks(componentName: string): ToolboxItems.ContentsType[] {
  return [
    // Method: set_position(pos: float) -> None
    {
      kind: 'block',
      type: 'mrc_call_python_function',
      extraState: {
        functionKind: 'instance_component',
        returnType: 'None',
        args: [
          {
            name: 'pos',
            type: 'float',
          },
        ],
        tooltip: 'Set the servo to a position between 0 and 1.',
        importModule: '',
        componentClassName: 'rev.Servo',
        componentName,
      },
      fields: {
        COMPONENT_NAME: componentName,
        FUNC: 'set_position',
      },
      inputs: {
        ARG0: {
          block: {
            type: 'math_number',
            fields: {
              NUM: 0.5,
            },
          },
        },
      },
    },
    // Method: set_angle_degrees(angle: float) -> None
    {
      kind: 'block',
      type: 'mrc_call_python_function',
      extraState: {
        functionKind: 'instance_component',
        returnType: 'None',
        args: [
          {
            name: 'angle',
            type: 'float',
          },
        ],
        tooltip: 'Set the servo to an angle between 0 and 270.',
        importModule: '',
        componentClassName: 'rev.Servo',
        componentName,
      },
      fields: {
        COMPONENT_NAME: componentName,
        FUNC: 'set_angle_degrees',
      },
      inputs: {
        ARG0: {
          block: {
            type: 'math_number',
            fields: {
              NUM: 180,
            },
          },
        },
      },
    },
  ];
}
